hi-ml-azure/src/health/azure/azure_util.py

```diff
@@ -523,3 +526,72 @@ def get_aml_runs(args: Namespace, workspace: Workspace, run_id_source: AzureRunI
     else:
         raise ValueError(f"Unrecognised RunIdSource: {run_id_source}")
     return [run for run in runs if run is not None]
+
+
+def hash_file(filename: str) -> str:
+    """
+    Compute a SHA1 hash of a file in 64kb chunks.
+
+    :param filename: File to hash.
+    :return: Hex version of SHA1 digest.
+    """
+    BLOCKSIZE = 65536
+    sha1 = hashlib.sha1()
+    with open(filename, 'rb') as afile:
+        while True:
+            data = afile.read(BLOCKSIZE)
+            if not data:
+                break
+            sha1.update(data)
+    return sha1.hexdigest()
+
+
+def run_upload_folder(run: Run,
+                      name: str,
+                      path: str,
+                      datastore_name: Optional[str] = None) -> Union[Tuple[dict, map], map]:
+    """
+    Wrap a call to run.upload_folder with extra checks to see if files already exist.
+
+    :param run: AzureML run to upload folder to.
+    :param name: The name of the folder of files to upload.
+    :param path: The relative local path to the folder to upload.
+    :param datastore_name: Optional DataStore name
+    """
+    # Get list of files already uploaded to the run wiith this name
+    existing_file_names = {f for f in run.get_file_names() if f.startswith(f"{name}/")}
+    # Get list of files in the local folder
+    local_files = {f for f in Path(path).rglob("*") if f.is_file()}
+    # Get list of file names as they would be after upload
+    local_file_named = {(f"{name}/{str(f.relative_to(path))}", str(f), f.name) for f in local_files}
+    # Filter out the files that are both local and already uploaded
+    dup_files = [f for f in local_file_named if f[0] in existing_file_names]
+    # Filter out the files that are local but not uploaded.
+    new_files = [f for f in local_file_named if f[0] not in existing_file_names]
+
+    # If there are duplicate files then upload_folder fails
+    if len(dup_files) == 0:
+        # If there are no duplicate files, then use upload_folder
+        return run.upload_folder(name=name, path=path, datastore_name=datastore_name)
+    else:
+        # Check the duplicate files.
+        # with TemporaryDirectory() as d:
+        d = Path("outputs") / "test_download_folder"
+        if d.exists():
+            shutil.rmtree(d)
+        d.mkdir()
+
+        for f in dup_files:
+            run.download_file(name=f[0],
+                              output_file_path=d)
+            downloaded_file = d / f[2]
+            downloaded_file_hash = hash_file(str(downloaded_file))
+
+            local_file_hash = hash_file(f[1])
+            if downloaded_file_hash != local_file_hash:
+                raise Exception(f"Files are different, {f[1]}")
```

Comment:
Can you say more in the exception? something along "trying to upload file ... but that file already exists in the run. The existing file on the run has hash..., but the local file has hash ... Unable to reconcile those differences"